### PR TITLE
Fixed #21408 — German Translation for “3 days ago”

### PR DESCRIPTION
**Pull request by claudep**
The problem:
“3 days ago” should translate to “vor 3 Tagen” in German, while “3 days” translates to “3 Tage”. #21408 describes that django always translated to “Tage”, even when the dative “Tagen” was correct. The same applies to months (“Monate”/“Monaten”) and years (“Jahre”/“Jahren”).

The solution:
Let `timesince` caller provide the string dict to use for the time-related strings.

### CODE REVIEW

**Comment by claudep** [comment]
Things to think about:
 - Are 'naturaltime-past'/'naturaltime-future' good enough to give translators a good hint (could be e.g. '[time] ago]'/'[time] from now'). Or we could add a translator comment in addition of the context.
 - Do '%(delta)s ago'/'%(delta)s from now' strings need to have a context?

File: django/utils/timesince.py [hunk 1]
```diff
@@ -26,9 +35,15 @@ def timesince(d, now=None, reversed=False):
     displayed.  For example, "2 weeks, 3 days" and "1 year, 3 months" are
     possible outputs, but "2 weeks, 3 hours" and "1 year, 5 days" are not.
 
+    `time_strings` is an optional dict of strings to replace the default
```
Comment (codingjoe) [inline django/utils/timesince.py]: I don't see a use case for the new argument right now. If you want to replace the strings, it seems more commen to provide your own PO gettext files/translations. That way you can provide your own strings, while not changing the place holders.

File: django/utils/timesince.py [hunk 2]
```diff
@@ -59,18 +74,18 @@ def timesince(d, now=None, reversed=False):
...
         seconds2, name2 = TIMESINCE_CHUNKS[i + 1]
         count2 = (since - (seconds * count)) // seconds2
         if count2 != 0:
-            result += gettext(', ') + avoid_wrapping(name2 % count2)
+            result += gettext(', ') + avoid_wrapping(time_strings[name2] % count2)
     return result
 
 
-def timeuntil(d, now=None):
+def timeuntil(d, now=None, time_strings=None):
```
Comment (codingjoe) [inline django/utils/timesince.py]: Same here, it may just drive complexity.

File: tests/humanize_tests/tests.py [hunk 3]
```diff
@@ -94,6 +94,32 @@ def test_intcomma_without_number_grouping(self):
...
+            'vor 2\xa0Tagen',
+            'vor 1\xa0Monat',
+            'vor 2\xa0Monaten',
+            'vor 1\xa0Jahr, 4\xa0Monaten',
+            'vor 2\xa0Jahren, 4\xa0Monaten',
+        ]
+
+        orig_humanize_datetime, humanize.datetime = humanize.datetime, MockDateTime
+        try:
+            with translation.override('de'), self.settings(USE_L10N=True):
```
Comment (codingjoe) [inline tests/humanize_tests/tests.py]: I know you didn't write the self.humanize_test, but I would suggest to use subtests here:
```python
with self.subTest(i)
```

Comment (codingjoe) [reply tests/humanize_tests/tests.py]: See https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

File: django/utils/timesince.py (hunk 1 above)
Comment (codingjoe) [reply django/utils/timesince.py]: aaah, ok, now I get it.

File: tests/humanize_tests/tests.py (hunk 3 above)
Comment (claudep) [reply tests/humanize_tests/tests.py]: Using `subTest` in `HumanizeTests.humanize_tester` should be the subject of a different patch.

Comment (codingjoe) [reply tests/humanize_tests/tests.py]: True, just thought I mention it.

**Comment by codingjoe** [comment]
No further comments, LGTM.

**Comment by claudep** [comment]
Removing context from '%(delta)s ago'/'%(delta)s from now' strings is addressed in #9930.

**Comment by claudep** [comment]
The test failure is unrelated.

**Comment by claudep** [comment]
Was merged to https://github.com/django/django/commit/78912ccd0e1fcdfe98ca85081c9eb8acb7aa1f6d

**Pull request closed without merging**